File: src/main/scala/org/apache/spark/ui/SparkCubePage.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.spark.ui

import javax.servlet.http.HttpServletRequest

import scala.xml.Node

import com.alibaba.sparkcube.optimizer.CacheIdentifier

// UIUtils is private to spark
case class SparkCubeInformation(
    viewDataBase: String,
    viewName: String,
    cacheName: String,
    enableRewrite: Boolean,
    cacheType: String,
    cacheDataSize: String,
    lastUpdateTime: Long,
    detailUrl: String,
    fullCacheName: String)

class SparkCubePage(parent: SparkCubeTab)
  extends AbstractSparkCubePage(parent, "") {

  override def render(request: HttpServletRequest): Seq[Node] = {

    val cacheManager = parent.sharedState.cubeManager
    val action = UIUtils.stripXSS(request.getParameter("action"))
    if (action != null && action.nonEmpty) {
      if (action == "drop") {
        val parameterCacheId = UIUtils.stripXSS(request.getParameter("cacheId"))
        require(parameterCacheId != null && parameterCacheId.nonEmpty, "Missing cache id parameter")
        val cacheId = CacheIdentifier(parameterCacheId)
        cacheManager.dropCache(sparkSession, cacheId)
      }
    }

    val caches = cacheManager.listAllCaches(sparkSession)
    var cachedDataInfo = Seq[SparkCubeInformation]()
    for(cache <- caches) {
      val viewDataBase = cache._1.database.getOrElse("")
      val viewName = cache._1.table
      var cacheName: String = null
      var enableRewrite: Boolean = false
      var cacheType: String = null
      var cacheDataSize: String = ""
      var lastUpdateTime: Long = -1
      var detailUrl = ""
      if (cache._2.rawCacheInfo.isDefined) {
        cacheName = cache._2.rawCacheInfo.get.cacheName
        detailUrl = jobURL(request, s"$viewDataBase.$viewName.$cacheName")
        enableRewrite = cache._2.rawCacheInfo.get.enableRewrite
        cacheType = "Raw Cache"
        cacheDataSize = getFileSize(cache._2.rawCacheInfo.get.storageInfo.storagePath)
        lastUpdateTime = cache._2.rawCacheInfo.get.lastUpdateTime
        val tempRelation = SparkCubeInformation(
          viewDataBase, viewName, cacheName, enableRewrite,
          cacheType, cacheDataSize, lastUpdateTime, detailUrl,
          s"$viewDataBase.$viewName.$cacheName")
        cachedDataInfo = cachedDataInfo :+ tempRelation
      }

      if (cache._2.cubeCacheInfo.isDefined) {
        cacheName = cache._2.cubeCacheInfo.get.cacheName
        detailUrl = jobURL(request, s"$viewDataBase.$viewName.$cacheName")
        enableRewrite = cache._2.cubeCacheInfo.get.enableRewrite
        cacheType = "Cube Cache"
        cacheDataSize = getFileSize(cache._2.cubeCacheInfo.get.storageInfo.storagePath)
        lastUpdateTime = cache._2.cubeCacheInfo.get.lastUpdateTime
        val tempRelation = SparkCubeInformation(
          viewDataBase, viewName, cacheName, enableRewrite,
          cacheType, cacheDataSize, lastUpdateTime, detailUrl,
          s"$viewDataBase.$viewName.$cacheName")
        cachedDataInfo = cachedDataInfo :+ tempRelation
      }
    }
    val content = generateLoadResources(request) ++ cacheTable(cachedDataInfo) ++
      addCreateButtonHtml(request) :+
      <div id="actionModal" class="modal hide" tabindex="-1" role="dialog"
           aria-labelledby="actionTitle" aria-hidden="true">
        <div class="modal-dialog" role="document">
          <div class="modal-content">
            <div class="modal-header">
              <h5 class="modal-title" id="actionTitle">

              </h5>
            </div>
            <div class="modal-body">
              <p id="actionBody">

              </p>
            </div>
            <div class="modal-footer">
              <a href={cubeUrl(request)} class="btn btn-primary">Close</a>
            </div>
          </div>
        </div>
      </div>
    UIUtils.headerSparkPage(request, "Cube Management", content, parent)
  }

  private def addCreateButtonHtml(request: HttpServletRequest): Seq[Node] = {
    <div>
      <a href={s"${UIUtils.prependBaseUri(request, parent.basePath)}/${parent.prefix}/Create"}
         class="btn btn-default">
        New Cache
      </a>
    </div>
  }

  private def cacheTable(rdds: Seq[SparkCubeInformation]): Seq[Node] = {
    if (rdds.isEmpty) {
      // Don't show the rdd table if there is no RDD persisted.
      Nil
    } else {
      <div>
        {UIUtils.listingTable(rcHeader, cacheRow, rdds, id = Some("cache-table"))}
      </div>
    }
  }

  /** Render an HTML row representing an Cube Management */
  private def cacheRow(cache: SparkCubeInformation): Seq[Node] = {
    val onclickFunc = "dropCache('" + {cache.fullCacheName} + "')"
    // scalastyle:off
    <tr>
      <td>{cache.viewDataBase}</td>
      <td>{cache.viewName}</td>
      <td>{cache.cacheName}</td>
      <td>{cache.enableRewrite}</td>
      <td>{cache.cacheType}</td>
      <td>{cache.cacheDataSize}</td>
      <td>{UIUtils.formatDate(cache.lastUpdateTime)}</td>
      <td><a href={cache.detailUrl}>{"Detail"}</a>
        <span>&nbsp;</span>
        <button class="btn btn-link" type="button" onclick={onclickFunc}>Drop</button>
      </td>
    </tr>
    // scalastyle:on
  }

  /** Header fields for the Rational Cache */
  private val rcHeader = Seq(
    "viewDataBase",
    "viewName",
    "cacheName",
    "enableRewrite",
    "cacheType",
    "cacheDataSize",
    "lastUpdataTime",
    "action"
  )

  private def jobURL(request: HttpServletRequest, cacheId: String): String =
    "%s/Cube Management/Detail/?cacheId=%s"
      .format(UIUtils.prependBaseUri(request, parent.basePath), cacheId)

  private def dropUrl(request: HttpServletRequest, cacheId: String): String = {
    "%s/Cube Management/?action=drop&cacheId=%s"
      .format(UIUtils.prependBaseUri(request, parent.basePath), cacheId)
  }

  private def cubeUrl(request: HttpServletRequest): String = {
    "%s/Cube Management"
      .format(UIUtils.prependBaseUri(request, parent.basePath))
  }

  private def generateLoadResources(request: HttpServletRequest): Seq[Node] = {
    // scalastyle:off
      <script src={UIUtils.prependBaseUri(request, "/static/bootstrap.min.js")}></script>
      <script src={UIUtils.prependBaseUri(request,
        "/static/SparkCube/caching/spark-cube-page.js")}></script>
    // scalastyle:on
  }
}

File: src/main/resources/com/alibaba/sparkcube/execution/ui/static/caching/spark-cube-page.js
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the 'License'); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

$(function() {
    $("span.additional-dimension").click(function() {
        $(this).parent().find('input[type="checkbox"]').trigger('click');
    });

    $("#select-all-dimension").click(function() {
       var status = window.localStorage.getItem("select-all-dimension") == "true";
       status = !status;
       if (this.checked) {
          // Toggle all un-checked options.
          $('input[name="cacheField"][type="checkbox"]:not(:checked)').trigger('click');
       } else {
          // Toggle all checked options.
          $('input[name="cacheField"][type="checkbox"]:checked').trigger('click');
       }
       window.localStorage.setItem("select-all-dimension", "" + status);
    });

    $("#close-cache").click(function(){
       $("#mymodal-data").toggle();
       var url = window.location.protocol + '//' + window.location.host + window.location.pathname;
       window.location.replace(url);
    });

    $("#create-cache").click(function(){
       var cacheType = $.trim($('#cacheType').text());
       var db = $.trim($('#cacheDataBase').text());
       var table = $.trim($('#cacheTableName').text());
       var cacheFieldArray = [];
       $.each($('input[name="cacheField"]:checkbox:checked'),function(){
         cacheFieldArray.push($(this).val());
       });
       var cacheRewrite = $('#cacheRewrite option:selected').text()
       var cacheProvider = $('#cacheProvider option:selected').text()
       var cachePartitionColumnArray = [];
       $.each($('input[name="cachePartitionColumn"]:checkbox:checked'),function(){
         cachePartitionColumnArray.push($(this).val());
       });
       var ZOrderColumnArray = [];
       $.each($('input[name="cacheZOrderColumn"]:checkbox:checked'),function(){
         ZOrderColumnArray.push($(this).val());
       });
       var cache = $('#cacheName').val();
       const createPath = "rcApi/v1/caches/" + db + "." + table;
       var url = location.protocol + '//' +
        location.host +
        location.pathname.replace(/(Cube%20Management\/Create)/, createPath) +
        location.search;
       var cacheDetail = {
           cacheName: cache,
           rewriteEnabled: cacheRewrite,
           provider: cacheProvider,
           partitionColumns: cachePartitionColumnArray,
           zorderColumns: ZOrderColumnArray
       }
       var flag = false;
       var p = new RegExp('^[a-zA-Z][a-zA-Z0-9_-]{0,19}$');
       if (cacheName.length == 0 || !p.test(cache)) {
          alert("Invalid cache name")
       } else if (cacheFieldArray.length ==0) {
          alert("please choose field")
       } else if (!judgeSubset(cachePartitionColumnArray, cacheFieldArray)) {
          alert("the partition fields should within the scope of dimension fields")
       } else if (!judgeSubset(ZOrderColumnArray, cacheFieldArray)) {
          alert("the zOrder fields should within the scope of dimension fields")
       } else if (judgeIntersect(cachePartitionColumnArray, ZOrderColumnArray)) {
          alert("the partition and zOrder field can't be same")
       } else if (cacheType == "RawCache") {
          cacheDetail.type = "raw"
          cacheDetail.schema = {
            selectColumns: cacheFieldArray
          }
          flag = true;
       } else if (cacheType == 'CubeCache') {
           var measureFieldArray = []
           $.each($('select[name="measureField"] option:checked'),function(){
              measureFieldArray.push($(this).val());
           });
           var measureFunctionArray = []
           $.each($('select[name="measureFunction"] option:checked'),function(){
             measureFunctionArray.push($(this).val());
           });
           if (measureFieldArray.filter(v => v.length == 0).length > 0) {
             alert("please choose measure field");
           } else if (measureFunctionArray.filter(v => v.length == 0).length > 0) {
             alert("please choose measure function");
           } else {
             for (var i=0; i < measureFunctionArray.length; i++) {
                if (measureFunctionArray[i] == "AVERAGE") {
                  measureFunctionArray.pop(measureFunctionArray[i]);
                  var field = measureFieldArray[i];
                  measureFieldArray.pop(field);
                  measureFieldArray.push(field);
                  measureFieldArray.push(field);
                  measureFunctionArray.push("COUNT");
                  measureFunctionArray.push("SUM");
                }
             }
             cacheDetail.type = "cube"
             cacheDetail.schema = {
               dimensions: cacheFieldArray,
               measuresFiled: measureFieldArray,
               measuresFunction: measureFunctionArray
             }
             flag = true;
           }
       }

       if (flag == true) {
           $.ajax({
                url: url,
                type: "PUT",
                contentType: "application/json",
                dataType: "json",
                data: JSON.stringify({param: JSON.stringify(cacheDetail)}),
                headers: {
                    "Accept": "application/json",
                    "Content-Type": "application/json",
                    "X-HTTP-Method-Override": "PUT" },
                success: function(response) {
                    $('#mymodal-data').toggle();
                    if (response.status == 'SUCCEED') {
                      displayModal("Create Cache",
                      "Create cache " + table +" successfully")
                    } else if (response.status == 'FAILED'){
                      displayModal("Create Cube Management",
                      "Cache already exists for " + table)
                    } else {
                      displayModal("Create Cache",
                      response.message)
                    }
                }
           })
       }
    });
})

function judgeSubset(a, b) {
    for (var i=0; i<a.length; i++) {
        if (b.indexOf(a[i]) == -1) {
            return false;
        }
    }
    return true;
}

function judgeIntersect(a, b) {
    for (var i=0; i<a.length; i++) {
        if (b.indexOf(a[i]) != -1) {
            return true;
        }
    }
    return false;
}

function dropCache(cacheId) {
  const dropPath = "rcApi/v1/caches/" + cacheId + "/delete";
  const url = location.protocol + '//' + location.host +
    location.pathname.replace(/(Cube%20Management)/, dropPath) + location.search;
  $.ajax({
    url: url,
    type: "PUT",
    contentType: "application/json",
    dataType: "json",
    headers: {
      "Accept": "application/json",
      "Content-Type": "application/json",
      "X-HTTP-Method-Override": "PUT" },
      success: function(response) {
        if (response.status == 'SUCCEED') {
          displayModal("Drop Cache",
            "Succeed to drop Cache " + cacheId + ".");
        } else {
          displayModal("Drop Cache",
            "Failed to drop cache " + cacheId +
             ".\n" + response.message);
        }
      }
  })
}

function deleteBuildFilter() {
  const div = this.parentNode;
  div.parentNode.removeChild(div);
}

function insertAfter( newElement, targetElement ) {
  var parent = targetElement.parentNode;
  if( parent.lastChild == targetElement ){
    parent.appendChild(newElement);
  }else{
    parent.insertBefore(newElement, targetElement.nextSibling);
  }
}

function addMeasure() {
  var nodeAll = document.getElementsByClassName('additional-metrics example')
  var node = nodeAll[nodeAll.length - 1]
  insertAfter(node.cloneNode(true), node)
}

function removeClick(obj) {
  if ($(".additional-metrics.example").length == 1) {
    alert("one measure field and function at least")
  } else {
    obj.parentElement.remove()
  }
}

function displayModal(title, bodyText) {
  $('#actionTitle').html(title)
  $('#actionBody').html(bodyText)
  $('#actionModal').modal('show')
}
